operator/cmd/k8s_identity.go

```diff
@@ -10,78 +10,208 @@ import (
+		identityResource: deps.IdentityResource,
+		heartbeatStore: identity.NewIdentityHeartbeatStore(
+			cfg.IdentityHeartbeatTimeout,
+		),
+		rateLimiter: rate.NewLimiter(
+			cfg.RateInterval,
+			cfg.RateLimit,
+		),
 	}
 
-	return err
+	deps.LC.Append(obj)
+
+	return obj
 }
 
-func updateIdentity(ctx context.Context, clientset k8sClient.Clientset, identity *v2.CiliumIdentity) error {
-	_, err := clientset.CiliumV2().CiliumIdentities().Update(
-		ctx,
-		identity,
-		metav1.UpdateOptions{})
-	if err != nil {
-		log.WithError(err).Error("Updating Identity")
-	} else {
-		log.WithField(logfields.Identity, identity.GetName()).Debug("Updated identity")
+// IdentitiesGC represents the Cilium identities periodic GC.
+type IdentitiesGC struct {
+	logger logrus.FieldLogger
+
+	cfg IdentitiesGCConfig
+
+	clientset        k8sClient.Clientset
+	identityResource resource.Resource[*v2.CiliumIdentity]
+
+	heartbeatStore *identity.IdentityHeartbeatStore
+	mgr            *controller.Manager
+	rateLimiter    *rate.Limiter
+
+	// counters for GC failed/successful runs
+	failedRuns     float64
+	successfulRuns float64
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Start implements hive.HookInterface
+func (igc *IdentitiesGC) Start(ctx hive.HookContext) error {
+	if igc.cfg.AllocationMode != option.IdentityAllocationModeCRD {
+		return nil
+	}
+
+	if !igc.clientset.IsEnabled() {
+		igc.logger.Fatal("CRD Identity allocation mode requires k8s to be configured.")
+	}
+
+	if igc.cfg.EndpointGCInterval == 0 {
+		log.Fatal("The CiliumIdentity garbage collector requires the CiliumEndpoint garbage collector to be enabled")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	igc.cancel = cancel
+
+	// start heartbeat updater
+	igc.wg.Add(1)
+	go igc.runHeartbeatUpdater(ctx)
+
+	// start identities gc
+	if igc.cfg.IdentityGCInterval != 0 {
+		igc.logger.
+			WithField(logfields.Interval, igc.cfg.IdentityGCInterval).
+			Info("Starting CRD identity garbage collector")
+
+		igc.mgr = controller.NewManager()
+		igc.mgr.UpdateController("crd-identity-gc",
+			controller.ControllerParams{
+				RunInterval: igc.cfg.IdentityGCInterval,
+				DoFunc: func(ctx context.Context) error {
+					igc.gc(ctx)
+					return ctx.Err()
+				},
+			})
 	}
 
-	return err
+	return nil
 }
 
-var identityHeartbeat *identity.IdentityHeartbeatStore
+// Stop implements hive.HookInterface
+func (igc *IdentitiesGC) Stop(ctx hive.HookContext) error {
+	// stop GC
+	igc.mgr.RemoveAllAndWait()
 
-// counters for GC failed/successful runs
-var (
-	failedRuns     = 0
-	successfulRuns = 0
-)
+	// stop heartbeat updater
+	igc.cancel()
+
+	igc.rateLimiter.Stop()
 
-// identityGCIteration is a single iteration of a garbage collection. It will
+	// wait for cleanup
+	igc.wg.Wait()
+
+	return nil
+}
+
+func (igc *IdentitiesGC) runHeartbeatUpdater(ctx context.Context) {
+	defer igc.wg.Done()
+
+	events := igc.identityResource.Events(ctx)
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
No need for this case. The events channel closes when ctx is cancelled, e.g. do:
`for event := range events {` instead.